sdl-debezium/src/main/scala/io/smartdatalake/workflow/dataobject/DebeziumCdcDataObject.scala

```diff
@@ -0,0 +1,329 @@
+    props = props ++ Map("table.include.list" -> table.fullName)
+
+    val propsForEngine = new Properties();
+    props.foreach { case (key, value) => propsForEngine.setProperty(key, value) }
+    propsForEngine
+  }
+
+  override def factory: FromConfigFactory[DataObject] = DebeziumCdcDataObject
+
+  override def getSparkDataFrame(partitionValues: Seq[PartitionValues])(implicit context: ActionPipelineContext): DataFrame = {
+
+    val spark = context.sparkSession
+
+    def getRecordsFromDebeziumEngine(
+                           properties: Properties,
+                           changeConsumer: DebeziumEngine.ChangeConsumer[ChangeEvent[SourceRecord, SourceRecord]] with HasRecords[SourceRecord],
+                           executorService: ExecutorService = Executors.newSingleThreadExecutor,
+                           timeoutMilliSeconds: Int = 10000
+                         ): Seq[SourceRecord] = {
+
+      val completionCallback = new DebeziumCompletionCallback(executorService)
+
+      val engine = DebeziumEngine.create(classOf[Connect])
+        .using(properties)
+        .notifying(changeConsumer)
+        .using(completionCallback)
+        .build()
+
+      executorService.execute(engine)
+
+      Thread.sleep(timeoutMilliSeconds)
+
+      engine.close()
+      executorService.shutdown()
+
+      completionCallback.error.foreach(err => throw new Exception(err))
+
+      changeConsumer.records
+
+    }
+
+    def createEmptyDataFrame(): DataFrame = {
+
+        val schemaProperties = debeziumPropertiesForEngine
+
+        Seq("offset.storage", "offset.storage.sdlb.data.object.id").foreach(schemaProperties.remove(_))
+
+        schemaProperties.put("offset.storage", "org.apache.kafka.connect.storage.MemoryOffsetBackingStore")
+
+        val records = getRecordsFromDebeziumEngine(schemaProperties, changeConsumer = new DebeziumSchemaConsumer)
+
+        val df = DebeziumEventConverter.convert(records)(spark)
+
+        val schema = df.schema
+
+        spark.createDataFrame(new util.ArrayList[Row](), schema)
+
+    }
+
+    if (context.isExecPhase) {
+
+      val records = getRecordsFromDebeziumEngine(debeziumPropertiesForEngine, changeConsumer = new DebeziumChangeConsumer)
+
+      records.headOption match {
+        case Some(_) => {
+          DebeziumEventConverter.convert(records)(spark)
+        }
+        case None => createEmptyDataFrame()
+      }
+
+    } else {
+      createEmptyDataFrame()
+    }
+
+  }
+
+  private[smartdatalake] var incrementalState: mutable.Map[String, String] = mutable.Map()
+
+  /**
+   * To implement incremental processing this function is called to initialize the DataObject with its state from the last increment.
+   * The state is just a string. It's semantics is internal to the DataObject.
+   * Note that this method is called on initializiation of the SmartDataLakeBuilder job (init Phase) and for streaming execution after every execution of an Action involving this DataObject (postExec).
```

Comment:
with streaming do you mean actual spark streaming here?

Reply:
I copied from another class from zach and me where we use this. The "streaming" is actually the sdlb streaming mode. Not spark streaming.

Reply:
Then I suggest this:
```suggestion
   * Note that this method is called on initializiation of the SmartDataLakeBuilder job (init Phase). When starting SDLB with the --streaming option, if will be called after every execution of an Action involving this DataObject (postExec).
```
Actually, the streaming Option should be renamed to micro batching cos that's what this is, but this is for another time...